### PR TITLE
New enos_facts, + module_utils/enos.py. modifying copyright year in rest all

### PR DESCRIPTION
**Pull request by amuraleedhar**
##### SUMMARY
Creating the modules for enos from Lenovo.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/network/enos/enos_facts.py
lib/ansible/module_utils/enos.py
lib/ansible/plugins/action/enos_facts.py
lib/ansible/plugins/cliconf/enos.py
lib/ansible/plugins/terminal/enos.py

##### ANSIBLE VERSION
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible-2.5.0-py2.7.egg/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

##### ADDITIONAL INFORMATION

This is an effort to create modules for enos based switches from Lenovo.

### CODE REVIEW

File: lib/ansible/module_utils/enos.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    'context': dict(),
+    'passwords': dict()
+}
+
+enos_argument_spec = {
+    'provider': dict(type='dict', options=enos_provider_spec),
+}
+
+enos_top_spec = {
+    'host': dict(removed_in_version=2.9),
```
Comment (gundalow) [inline lib/ansible/module_utils/enos.py]: You can remove the top-level arguments that you have `removed_in_version=2.9`. No point adding something to remove it in the future.

People (& examples) should be using `provider:`

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        with an initial C(M(!)) to specify that a specific subset should
+        not be collected.
+    required: false
+    default: '!config'
+'''
+EXAMPLES = '''
+Tasks: The following are examples of using the module enos_facts.
+---
+- name: Test Enos Facts
+  enos_facts: provider={{ cli }}
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py]: Please use multi-line yaml

```
- name: Test Enos Facts
  enos_facts:
    provider: "{{ cli }}"
...
```

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+      port: 22
+      username: admin
+      password: admin
+      transport: cli
+      timeout: 30
+      authorize: True
+      auth_pass:
+
+---
+# Collect all facts from the device
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py:77]: Please use `-name: ` rather than comments

**Comment by dkasberg** [comment]
Can this be merged now?

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Module to Collect facts from Lenovo Switches running lenovo ENOS commands
+# Lenovo Networking
+#
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'network'}
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py]: I believe this should be community

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+
+DOCUMENTATION = '''
+---
+module: enos_facts
+version_added: "2.5"
+author: "Anil Kumar Muraleedharan (Anil)"
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py]: `This should be your GitHub ID, 

`author: "Anil Kumar Muraleedharan (@amuraleedhar)

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+    default: '!config'
+'''
+EXAMPLES = '''
+Tasks: The following are examples of using the module enos_facts.
+---
+- name: Test Enos Facts
+  enos_facts:
+    provider={{ cli }}
+  with_items:
+    "{{enos_facts_data}}"
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py]: Where is `enos_facts_data` defined?

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+    description: The configured hostname of the device
+    returned: always
+    type: string
+  ansible_net_image:
+    description: Indicates the active image for the device
+    returned: always
+    type: string
+# hardware
+  ansible_net_memfree_mb:
+    description: The available free memory on the remote device in Mb
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py]: `Mb` or `MB`?

Comment (gundalow) [reply lib/ansible/modules/network/enos/enos_facts.py]: This applies to a few places.

**Comment by gundalow** [comment]
For future reference changes like this should be broken down into two PRs
1) One for the module
2) Another for bulk changes (such as copyright)

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Module to Collect facts from Lenovo Switches running lenovo ENOS commands
```
Comment (gundalow) [inline lib/ansible/modules/network/enos/enos_facts.py]: `s/lenovo/Lenovo/`

**Comment by amuraleedhar** [comment]
Addressed all issues raised by gundalow. Can this be merged now?

**Comment by gundalow** [comment]
You need to squash commits and remove the `@user` from the commit message

**Comment by amuraleedhar** [comment]
How to update an older commit message?

**Comment by gundalow** [comment]
Backup your files
Then use Git to squash the 11 commits into 1

**Comment by amuraleedhar** [comment]
I think I have messed it up all. Will see tomorrow.

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+    COMMANDS = [
+        'show system memory'
+    ]
+
+    def populate(self):
+        super(Hardware, self).populate()
+        data = self.responses[0]
+        # file = open('Sunil.txt', "a")
+        # file.write(data)
+        # file.close()
```
Comment (ganeshrn) [inline lib/ansible/modules/network/enos/enos_facts.py]: This seems to debug code. Please remove the commented lines.

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+
+    def parse_image(self, data):
+        match = re.search(r'(.*) image1(.*)', data, re.M | re.I)
+        if match:
+            return "Image1"
+        else:
+            return "Image2"
+
+    def parse_serialnum(self, data):
+        # match = re.search(r'board ID (\S+)', data)
```
Comment (ganeshrn) [inline lib/ansible/modules/network/enos/enos_facts.py:224]: Same as above

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+            match = re.match(r'hostname (.*?)', line, re.M | re.I)
+            if match:
+                hosts = line.split()
+                hostname = hosts[1].strip('\"')
+                return hostname
+            else:
+                return "NA"
+
+    def parse_model(self, data):
+        # match = re.search(r'^Cisco (.+) \(revision', data, re.M)
```
Comment (ganeshrn) [inline lib/ansible/modules/network/enos/enos_facts.py]: Is this commented line required?

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+
+class FactsBase(object):
+
+    COMMANDS = list()
+
+    def __init__(self, module):
+        self.module = module
+        self.facts = dict()
+        self.responses = None
+        self.PERSISTENT_COMMAND_TIMEOUT = 60
```
Comment (ganeshrn) [inline lib/ansible/modules/network/enos/enos_facts.py:169]: I don't see `PERSISTENT_COMMAND_TIMEOUT` used anywhere. Is this required?

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+FACT_SUBSETS = dict(
+    default=Default,
+    hardware=Hardware,
+    interfaces=Interfaces,
+    config=Config,
+)
+
+VALID_SUBSETS = frozenset(FACT_SUBSETS.keys())
+
+PERSISTENT_COMMAND_TIMEOUT = 60
```
Comment (ganeshrn) [inline lib/ansible/modules/network/enos/enos_facts.py:417]: Same as above

File: lib/ansible/modules/network/enos/enos_facts.py [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+  enos_facts:
+    provider={{ cli }}
+
+  vars:
+    cli:
+      host: "{{ inventory_hostname }}"
+      port: 22
+      username: admin
+      password: admin
+      transport: cli
```
Comment (ganeshrn) [inline lib/ansible/modules/network/enos/enos_facts.py:71]: `transport` option is not present in `provider` spec in `module_uitls/enos.py`. Is this required?

**Review by ganeshrn — COMMENTED** [review]
Please add unit test for enos_facts module [here](https://github.com/ansible/ansible/tree/devel/test/units/modules/network)

**Comment by amuraleedhar** [comment]
Can this be merged now? Unit test cases I will add in another PR.

File: lib/ansible/modules/network/enos/enos_facts.py (hunk 9 above)
Comment (amuraleedhar) [reply lib/ansible/modules/network/enos/enos_facts.py]: Done

File: lib/ansible/modules/network/enos/enos_facts.py (hunk 10 above)
Comment (amuraleedhar) [reply lib/ansible/modules/network/enos/enos_facts.py:224]: Done

File: lib/ansible/modules/network/enos/enos_facts.py (hunk 11 above)
Comment (amuraleedhar) [reply lib/ansible/modules/network/enos/enos_facts.py]: Removed

File: lib/ansible/modules/network/enos/enos_facts.py (hunk 12 above)
Comment (amuraleedhar) [reply lib/ansible/modules/network/enos/enos_facts.py:169]: Removed. Its not required

File: lib/ansible/modules/network/enos/enos_facts.py (hunk 13 above)
Comment (amuraleedhar) [reply lib/ansible/modules/network/enos/enos_facts.py:417]: Removed this too

File: lib/ansible/modules/network/enos/enos_facts.py (hunk 14 above)
Comment (amuraleedhar) [reply lib/ansible/modules/network/enos/enos_facts.py:71]: Removed.

**Comment by gundalow** [comment]
Tests need adding in the same PR where the modules are added

**Comment by gundalow** [comment]
Shippable was failing with unrelated issues, so I've triggered a rerun

**Comment by amuraleedhar** [comment]
Please merge this now. 
Thanks

-Anil

**Review by gundalow — APPROVED** [review]

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged into `devel`, will be released in `2.5.0`

**Pull request merged**